mkosi/distributions/gentoo.py

```diff
@@ -11,188 +10,87 @@
 
 from mkosi.architecture import Architecture
 from mkosi.distributions import DistributionInstaller
-from mkosi.install import copy_path, flock
+from mkosi.install import copy_path
 from mkosi.log import ARG_DEBUG, complete_step, die, log_step
 from mkosi.remove import unlink_try_hard
 from mkosi.run import run, run_workspace_command
 from mkosi.state import MkosiState
+from mkosi.types import PathString
 
 
 def invoke_emerge(
     state: MkosiState,
+    sysroot: Path,
     pkgs: Sequence[str] = (),
     actions: Sequence[str] = (),
     opts: Sequence[str] = (),
+    env: dict[str, str] = {},
 ) -> None:
-    jobs = os.cpu_count() or 1
+    thread_counts = (os.cpu_count() or 1) * 2  # * 2 for hyperthreading
+    bwrap: list[PathString] = []
+    root = state.root
+    if sysroot:
+        # This is the mount-point inside our sysroot where we mount root
+        target_root_mntp = "/tmp/mkosi-root"
+        bwrap += ["--bind", state.root, target_root_mntp]
+        root = Path(target_root_mntp)
+    else:
+        sysroot = root
+
     emerge_default_opts = [
         "--buildpkg=y",
         "--usepkg=y",
         "--keep-going=y",
-        f"--jobs={jobs}",
-        f"--load-average={jobs+1}",
+        f"--jobs={thread_counts}",
+        f"--load-average={thread_counts+1}",
         "--nospinner",
+        *([f"--root={root}"] if root else []),
     ]
     if ARG_DEBUG.get():
         emerge_default_opts += ["--verbose", "--quiet=n", "--quiet-fail=n"]
     else:
         emerge_default_opts += ["--quiet-build", "--quiet"]
     cmd = ["emerge", *pkgs, *emerge_default_opts, *opts, *actions]
-    run_workspace_command(state.root, cmd, network=True, env=state.environment)
-
-
-class Gentoo:
-    arch_profile: Path
-    arch: str
-    custom_profile_path: Path
-    ebuild_sh_env_dir: Path
-    emerge_vars: dict[str, str]
-    portage_cfg_dir: Path
-    profile_path: Path
-    root: Path
-    pkgs: dict[str, list[str]] = {}
-
-    EMERGE_UPDATE_OPTS = [
-        "--update",
-        "--tree",
-        "--changed-use",
-        "--newuse",
-        "--deep",
-        "--with-bdeps=y",
-        "--complete-graph-if-new-use=y",
-        "--verbose-conflicts",
+    bwrap += [
+        "--bind", state.cache_dir / "binpkgs", "/var/cache/binpkgs",
+        "--bind", state.cache_dir / "distfiles", "/var/cache/distfiles",
+        "--bind", state.cache_dir / "repos", "/var/db/repos",
```

Comment:
yes, `emerge` looks up `DISTDIR` in `env` for `distfiles` for instance. but why would we want to set them to non-default locations?

this is to preserve the `binpkgs`, `repos` and source codes (`distfiles`) in consecutive calls, since they're expensive to get/build.

would you explain a bit more please?

Reply:
Yes my bad I misunderstood that this is actually required